File: src/improvement/meta_analysis.py
"""Meta-analysis — analyze convergence data to improve methodology.

Correlates methodology elements with convergence speed, proposes
methodology changes with evidence.
"""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass
class ConvergenceDataPoint:
    """Data from a single convergence run."""
    plan_name: str
    total_rounds: int
    total_findings: int
    duration_seconds: float
    phases_used: list[str] = field(default_factory=list)
    escalated: bool = False


@dataclass
class MethodologyInsight:
    """An insight from meta-analysis."""
    element: str
    correlation: str  # positive, negative, neutral
    evidence: str
    suggestion: str


def analyze_convergence_history(
    data_points: list[ConvergenceDataPoint],
) -> list[MethodologyInsight]:
    """Analyze convergence history for methodology improvement signals.

    Identifies:
    - Which phases correlate with faster convergence
    - Which patterns lead to escalation
    - Optimal round counts
    """
    insights = []

    if not data_points:
        return insights

    # Average rounds
    avg_rounds = sum(d.total_rounds for d in data_points) / len(data_points)

    # Escalation rate
    escalated = sum(1 for d in data_points if d.escalated)
    escalation_rate = escalated / len(data_points)

    if escalation_rate > 0.3:
        insights.append(MethodologyInsight(
            element="max_rounds",
            correlation="negative",
            evidence=f"{escalation_rate:.0%} of runs escalated",
            suggestion="Consider increasing max_rounds or breaking plans into smaller units",
        ))

    if avg_rounds < 2:
        insights.append(MethodologyInsight(
            element="plan_granularity",
            correlation="positive",
            evidence=f"Average {avg_rounds:.1f} rounds suggests plans are too simple",
            suggestion="Plans may not be challenging enough — add more audit dimensions",
        ))

    # Phase effectiveness
    phase_counts: dict[str, int] = {}
    for dp in data_points:
        for phase in dp.phases_used:
            phase_counts[phase] = phase_counts.get(phase, 0) + 1

    # Findings per round
    total_findings = sum(d.total_findings for d in data_points)
    total_rounds = sum(d.total_rounds for d in data_points)
    if total_rounds > 0:
        findings_per_round = total_findings / total_rounds
        if findings_per_round > 10:
            insights.append(MethodologyInsight(
                element="audit_thoroughness",
                correlation="positive",
                evidence=f"{findings_per_round:.1f} findings per round",
                suggestion="Audit dimensions are effective at finding issues",
            ))

    return insights

File: rust/src/improvement/meta_analysis.rs
//! Meta-analysis — analyze convergence data to improve methodology.

use serde::{Deserialize, Serialize};

/// Data from a single convergence run.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConvergenceDataPoint {
    pub plan_name: String,
    pub total_rounds: usize,
    pub total_findings: usize,
    pub duration_seconds: f64,
    #[serde(default)]
    pub phases_used: Vec<String>,
    #[serde(default)]
    pub escalated: bool,
}

/// An insight from meta-analysis.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MethodologyInsight {
    pub element: String,
    pub correlation: String,
    pub evidence: String,
    pub suggestion: String,
}

/// Analyze convergence history for methodology improvement signals.
pub fn analyze_convergence_history(
    data_points: &[ConvergenceDataPoint],
) -> Vec<MethodologyInsight> {
    let mut insights = Vec::new();

    if data_points.is_empty() {
        return insights;
    }

    let avg_rounds =
        data_points.iter().map(|d| d.total_rounds).sum::<usize>() as f64
            / data_points.len() as f64;

    let escalated = data_points.iter().filter(|d| d.escalated).count();
    let escalation_rate = escalated as f64 / data_points.len() as f64;

    if escalation_rate > 0.3 {
        insights.push(MethodologyInsight {
            element: "max_rounds".into(),
            correlation: "negative".into(),
            evidence: format!("{:.0}% of runs escalated", escalation_rate * 100.0),
            suggestion:
                "Consider increasing max_rounds or breaking plans into smaller units".into(),
        });
    }

    if avg_rounds < 2.0 {
        insights.push(MethodologyInsight {
            element: "plan_granularity".into(),
            correlation: "positive".into(),
            evidence: format!("Average {:.1} rounds suggests plans are too simple", avg_rounds),
            suggestion: "Plans may not be challenging enough — add more audit dimensions".into(),
        });
    }

    // Findings per round
    let total_findings: usize = data_points.iter().map(|d| d.total_findings).sum();
    let total_rounds: usize = data_points.iter().map(|d| d.total_rounds).sum();
    if total_rounds > 0 {
        let findings_per_round = total_findings as f64 / total_rounds as f64;
        if findings_per_round > 10.0 {
            insights.push(MethodologyInsight {
                element: "audit_thoroughness".into(),
                correlation: "positive".into(),
                evidence: format!("{:.1} findings per round", findings_per_round),
                suggestion: "Audit dimensions are effective at finding issues".into(),
            });
        }
    }

    insights
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_empty_data_points() {
        let insights = analyze_convergence_history(&[]);
        assert!(insights.is_empty());
    }

    #[test]
    fn test_high_escalation_rate() {
        let data = vec![
            ConvergenceDataPoint {
                plan_name: "plan1".into(),
                total_rounds: 5,
                total_findings: 10,
                duration_seconds: 100.0,
                phases_used: vec![],
                escalated: true,
            },
            ConvergenceDataPoint {
                plan_name: "plan2".into(),
                total_rounds: 5,
                total_findings: 8,
                duration_seconds: 80.0,
                phases_used: vec![],
                escalated: true,
            },
        ];
        let insights = analyze_convergence_history(&data);
        assert!(insights.iter().any(|i| i.element == "max_rounds"));
    }

    #[test]
    fn test_low_avg_rounds() {
        let data = vec![ConvergenceDataPoint {
            plan_name: "easy".into(),
            total_rounds: 1,
            total_findings: 0,
            duration_seconds: 10.0,
            phases_used: vec![],
            escalated: false,
        }];
        let insights = analyze_convergence_history(&data);
        assert!(insights.iter().any(|i| i.element == "plan_granularity"));
    }

    #[test]
    fn test_high_findings_per_round() {
        let data = vec![ConvergenceDataPoint {
            plan_name: "thorough".into(),
            total_rounds: 2,
            total_findings: 30,
            duration_seconds: 200.0,
            phases_used: vec![],
            escalated: false,
        }];
        let insights = analyze_convergence_history(&data);
        assert!(insights.iter().any(|i| i.element == "audit_thoroughness"));
    }
}
